optuna_train_custom_trainer.py

```diff
@@ -0,0 +1,143 @@
+        vocab: Vocabulary,
+        embedding_dim: int,
+        max_filter_size: int,
+        num_filters: int,
+        output_dim: int,
+        dropout: float,
+):
+    embedding = Embedding(embedding_dim=embedding_dim, trainable=True, vocab=vocab)
+    embedder = BasicTextFieldEmbedder({"tokens": embedding})
+    encoder = CnnEncoder(
+        ngram_filter_sizes=range(1, max_filter_size),
+        num_filters=num_filters,
+        embedding_dim=embedding_dim,
+        output_dim=output_dim,
+    )
+    model = BasicClassifier(
+        text_field_embedder=embedder,
+        seq2vec_encoder=encoder,
+        dropout=dropout,
+        vocab=vocab,
+    )
+    return model
+
+
+def objective_fn(trial, device: int, target_metric: str, base_serialization_dir: str):
+    embedding_dim = trial.suggest_int("embedding_dim", 16, 512)
+    max_filter_size = trial.suggest_int("max_filter_size", 3, 6)
+    num_filters = trial.suggest_int("num_filters", 16, 512)
+    output_dim = trial.suggest_int("output_dim", 16, 512)
+    dropout = trial.suggest_float("dropout", 0, 1.0)
+    lr = trial.suggest_float("lr", 1e-5, 1e-1, log=True)
+
+    train_dataset, valid_dataset, vocab = prepare_data()
+    model = create_model(vocab, embedding_dim, max_filter_size, num_filters, output_dim, dropout)
+
+    if device > -1:
+        model.to(torch.device("cuda:{}".format(device)))
+
+    optimizer = SGD(model.parameters(), lr=lr)
+    data_loader = DataLoader(train_dataset, batch_size=10, collate_fn=allennlp_collate)
+    validation_data_loader = DataLoader(valid_dataset, batch_size=64, collate_fn=allennlp_collate)
+    serialization_dir = os.path.join(base_serialization_dir, "trial_{}".format(trial.number))
+    trainer = GradientDescentTrainer(
+        model=model,
+        optimizer=optimizer,
+        data_loader=data_loader,
+        validation_data_loader=validation_data_loader,
+        validation_metric="+" + target_metric,  # TODO (himkt): use direction
+        patience=None,  # `patience=None` since it could conflict with AllenNLPPruningCallback
+        num_epochs=50,
+        cuda_device=device,
+        serialization_dir=serialization_dir,
+        epoch_callbacks=[AllenNLPPruningCallback(trial, "validation_" + target_metric)],
+    )
+    vocab.save_to_files(os.path.join(serialization_dir, "vocabulary"))
+    result = trainer.train()
+    return result[f"best_validation_{target_metric}"]
+
+
+@click.command()
+@click.option("--device", type=int, default=-1)
+@click.option("--target_metric", type=str, default="accuracy")
+@click.option("--base_serialization_dir", type=str, default="model")
+@click.option("--clear-at-end", is_flag=True)
+def main(device, target_metric, base_serialization_dir, clear_at_end):
+    os.makedirs(base_serialization_dir, exist_ok=True)
+    storage = "sqlite:///" + os.path.join(base_serialization_dir, "optuna.db")
+    study = optuna.create_study(
+        direction="maximize",
+        pruner=optuna.pruners.HyperbandPruner(),
```

Comment:
Use `Hyperband`, efficient bandit-based pruning algorithm ([[Li+, 2018]](http://jmlr.org/papers/volume18/16-558/16-558.pdf)).
Please see [Optuna documentation](https://optuna.readthedocs.io/en/stable/reference/pruners.html) for other available pruners